pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/LeaderService.java

```diff
@@ -0,0 +1,118 @@
+import org.apache.pulsar.client.api.ConsumerEventListener;
+import org.apache.pulsar.client.api.PulsarClient;
+import org.apache.pulsar.client.api.PulsarClientException;
+import org.apache.pulsar.client.api.SubscriptionType;
+import org.apache.pulsar.client.impl.ConsumerImpl;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@Slf4j
+public class LeaderService implements AutoCloseable, ConsumerEventListener {
+
+    private final String consumerName;
+    private final FunctionAssignmentTailer functionAssignmentTailer;
+    private final ErrorNotifier errorNotifier;
+    private ConsumerImpl<byte[]> consumer;
+    private final WorkerConfig workerConfig;
+    private final PulsarClient pulsarClient;
+    private final AtomicBoolean isLeader = new AtomicBoolean(false);
+
+    static final String COORDINATION_TOPIC_SUBSCRIPTION = "participants";
+
+    private static String WORKER_IDENTIFIER = "id";
+    
+    public LeaderService(WorkerService workerService,
+                         PulsarClient pulsarClient,
+                         FunctionAssignmentTailer functionAssignmentTailer,
+                         ErrorNotifier errorNotifier) {
+        this.workerConfig = workerService.getWorkerConfig();
+        this.pulsarClient = pulsarClient;
+        this.functionAssignmentTailer = functionAssignmentTailer;
+        this.errorNotifier = errorNotifier;
+        consumerName = String.format(
+                "%s:%s:%d",
+                workerConfig.getWorkerId(),
+                workerConfig.getWorkerHostname(),
+                workerConfig.getWorkerPort()
+        );
+
+    }
+
+    public void start() throws PulsarClientException {
+        // the leaders service is using a `coordination` topic for leader election.
+        // we don't produce any messages into this topic, we only use the `failover` subscription
+        // to elect an active consumer as the leader worker. The leader worker will be responsible
+        // for scheduling snapshots for FMT and doing task assignment.
+        consumer = (ConsumerImpl<byte[]>) pulsarClient.newConsumer()
+                .topic(workerConfig.getClusterCoordinationTopic())
+                .subscriptionName(COORDINATION_TOPIC_SUBSCRIPTION)
+                .subscriptionType(SubscriptionType.Failover)
+                .consumerEventListener(this)
+                .property(WORKER_IDENTIFIER, consumerName)
+                .subscribe();
+
+    }
+
+    @Override
+    public void becameActive(Consumer<?> consumer, int partitionId) {
+        if (isLeader.compareAndSet(false, true)) {
+            log.info("Worker {} became the leader.", consumerName);
+            try {
+                // trigger read to the end of the topic and exit
+                // Since the leader can just update its in memory assignments cache directly
+                functionAssignmentTailer.triggerReadToTheEndAndExit().get();
```

Comment:
This should be abstracted out from leaderservice to respective class(in this case functionruntimemanager)

Reply:
Also we need to create the producer here right?

Reply:
> This should be abstracted out from leaderservice to respective class(in this case functionruntimemanager)

yup done

> Also we need to create the producer here right?

Why do we need to create a producer?

Reply:
Yup. That is the same pattern in https://github.com/apache/pulsar/pull/7255 as well

Reply:
done